### PR TITLE
:sparkles: Add DLPack tensor export functionality to AIGs

### LINKED ISSUE #307: Feature Request: Zero-Copy  Tensors via DLPack
### What's the problem this feature will solve?

As discussed, loading massive AIGs into PyTorch Geometric (PyG) for deep learning requires an extremely efficient data pipeline. Copying flat arrays from C++ to Python and then again to PyTorch creates unnecessary memory overhead and heavily bottlenecks training loops when processing large datasets of circuits.

### Describe the solution you'd like

Being able to use `torch.from_dlpack()` allows for zero-copy data transfer directly into PyTorch—keeping the operation $O(1)$ regardless of graph size.

Following up on the machine learning integration thoughts in https://github.com/marcelwa/aigverse/issues/283, here is a breakdown of the specific DLPack capsules I'd need shaped and returned from the C++ side:

### 1. Graph Topology

Just need the non-zero entries for the edges:

* **`edge_index`**: Tensor of shape `(2, E)` representing the source and target nodes. *(Note: Must be `int64_t` so PyG doesn't force a memory copy to `torch.long`)*.

* **`edge_attr`**: Tensor of shape `(E, D_edge)`. *(Note: Allocate as `float32`)*.

  * Needs a flag to choose the type of encoding for the edge inversion (regular vs. inverted). Specifically:

    * **Integer `(0, 1)`**: Useful if passing into a categorical embedding layer in PyG.

    * **Integer `(1, -1)`**: Crucial if instantiating sparse matrix multiplication (using `0` would completely zero-out that edge type).

    * **One-hot `(E x 2)`**.

  * It would also be super helpful to explicitly document the mapping for these encodings (e.g., `0 = regular`, `1 = inverted`, or which index corresponds to which type in the one-hot vector) so there's no ambiguity on the Python side.

### 2. Node Attributes

* **`node_attr`**: Tensor of shape `(N, D_node)`. *(Note: Allocate as `float32`)*.

  * Needs a flag to choose the base encoding for node types (constant, PI, gate, PO): integer (base dim = 1) or one-hot encoded (base dim = `num_types` = 4).

  * **Optional Add-ons**: Boolean flags to concatenate additional features. Ideally, these should be stacked/concatenated along the feature dimension (axis 1) into a single contiguous block of memory on the C++ side. **`D_node`** will be the base dimension plus any included features:

    * `include_level=False`: Appends the logic level (+1 dimension).

    * `include_fanout=False`: Appends the number of fanouts (+1 dimension).

    * `include_truth_table=False`: Appends the local truth table per node (+`TT_dim` dimensions).

  * *(Example: If using one-hot, level, and fanout, they are stacked horizontally so `D_node` = 4 + 1 + 1 = 6).*

### Passing the Data via Dictionary

To pass these flat arrays back to Python, returning a dictionary of DLPack capsules seems like the gold standard for efficiency. The dictionary overhead is basically zero while keeping the Python API extremely clean and native to tools like PyTorch Geometric.

### Suggested Python Interaction

If the C++ binding (e.g., `to_graph_tensor`) returns a dictionary of these DLPack capsules, I can instantly build either a PyTorch Geometric graph or a native PyTorch sparse matrix on my end without any C++ overhead:

```python
import torch

# Call the C++ binding to get a dict of DLPack capsules
dlpack_data = aig.to_graph_tensor(
    node_encoding="one_hot", 
    edge_encoding="int_1_minus_1", # Specifying the safe integer format
    include_level=False,
    include_fanout=False
)

# Zero-copy conversion to PyTorch tensors!
edge_index = torch.from_dlpack(dlpack_data["edge_index"])   # Shape: (2, E)
edge_attr = torch.from_dlpack(dlpack_data["edge_attr"])     # Shape: (E, D_edge)
node_attr = torch.from_dlpack(dlpack_data["node_attr"])     # Shape: (N, D_node)

# If I need a mathematical Sparse Adjacency Matrix, I construct it instantly:
num_nodes = node_attr.shape[0]
sparse_adj = torch.sparse_coo_tensor(
    indices=edge_index, 
    values=edge_attr, 
    size=(num_nodes, num_nodes, edge_attr.shape[1])
)

```

As a quick thought, since building the sparse adjacency matrix this way is super handy but maybe not obvious to everyone, it might be worth adding a quick note or docstring to `to_graph_tensor` explaining this `torch.sparse_coo_tensor` trick. That way, anyone else who needs an adjacency matrix knows exactly how to build it efficiently from the sparse outputs!

### Example: PyTorch Geometric Dataset Integration

To show exactly why this format is so powerful, here is a quick mock-up of how this function lets us process a massive folder of AIGER files directly into PyTorch Geometric `Data` objects on the fly, keeping our RAM overhead tiny:

```python
from pathlib import Path
import torch
from torch_geometric.data import Dataset, Data
import aigverse

class AIGERDataset(Dataset):
    def __init__(self, root_dir):
        super().__init__(root_dir)
        # Find all .aig files in the target directory using pathlib
        self.file_paths = list(Path(root_dir).glob('*.aig'))

    def len(self):
        return len(self.file_paths)

    def get(self, idx):
        file_path = str(self.file_paths[idx])
        
        # 1. Parse AIG via aigverse
        aig = aigverse.read_aiger(file_path)
        
        # 2. Extract DLPack capsules using the proposed C++ function
        dlpack_data = aig.to_graph_tensor(
            node_encoding="one_hot", 
            edge_encoding="int_0_1", # Using 0/1 here since PyG Embeddings like standard class indices
            include_level=True,
            include_fanout=True
        )
        
        # 3. Zero-copy ingest into PyTorch
        x = torch.from_dlpack(dlpack_data["node_attr"])
        edge_index = torch.from_dlpack(dlpack_data["edge_index"])
        edge_attr = torch.from_dlpack(dlpack_data["edge_attr"])
        
        # 4. Return standard PyTorch Geometric Data object
        return Data(x=x, edge_index=edge_index, edge_attr=edge_attr)

# Usage: 
# dataset = AIGERDataset(root_dir="/path/to/massive/aig/folder")
# loader = DataLoader(dataset, batch_size=32, shuffle=True)
```

Would love to get your input on this!

### PR DESCRIPTION
**Pull request by marcelwa**
## Description

This pull request introduces a new feature for exporting AIG (And-Inverter Graph) networks as DLPack-compatible sparse tensors, enabling efficient, zero-copy integration with modern machine learning frameworks such as PyTorch, JAX, and TensorFlow. It adds new encoding options for node and edge attributes, updates documentation and Python/C++ interfaces, and adds PyTorch as a dedicated, test-only optional dependency used to exercise the DLPack interop tests.

**Major features and changes:**

#### DLPack Tensor Export for Machine Learning
* Adds a new `to_graph_tensors` method to the `Aig` class (Python and C++), allowing the export of graph topology and features as DLPack-compatible NumPy arrays, suitable for direct consumption by ML frameworks. The export supports configurable node and edge encodings, and optional inclusion of logic level, fanout, and truth-table features. [[1]](diffhunk://#diff-4d62a9db7d749f2ccd7ffbf13d7b09ddad06aa985fe467b950d962e27b25b218R309-R342) [[2]](diffhunk://#diff-6f8d201a6a0399da3354d935e58302f437007e3dc59041f9e3a3010b373c05aeR1-R278) [[3]](diffhunk://#diff-98009d538063c4d6228fe281f355afdc793908bb750e1c650c980937ff924b48R389-R425)

#### Node and Edge Encoding Options
* Introduces `NodeTensorEncoding` and `EdgeTensorEncoding` enums in both Python and C++ APIs, supporting integer and one-hot encodings for nodes, and binary, signed, or one-hot encodings for edges. These are exposed to Python and documented in both code and user documentation. [[1]](diffhunk://#diff-4d62a9db7d749f2ccd7ffbf13d7b09ddad06aa985fe467b950d962e27b25b218R7-R44) [[2]](diffhunk://#diff-6f8d201a6a0399da3354d935e58302f437007e3dc59041f9e3a3010b373c05aeR1-R278) Ff096675L113R113, [[3]](diffhunk://#diff-98009d538063c4d6228fe281f355afdc793908bb750e1c650c980937ff924b48R781)

#### Documentation and Examples
* Expands the user documentation with a new section on DLPack sparse tensor export, including practical end-to-end examples for PyTorch and NumPy, encoding/dtype mapping, and expected tensor shapes.

#### Dependency Updates
* The `adapters` code itself is tensor-framework-agnostic; PyTorch is only needed to exercise the DLPack interop tests. It is therefore its own `torch` dependency group in `pyproject.toml` (not part of the public `adapters` extra), and the 3 dependent tests are marked with `@pytest.mark.torch` and deselected by default (`-m "not torch"`). CI and cibuildwheel explicitly install the group and re-include the marked tests (`-m "torch or not torch"`), so coverage doesn't regress.

#### Internal C++ Implementation
* Implements the tensor export logic in a new C++ header (`graph_tensors.hpp`), handling the construction of edge and node feature arrays, DLPack-compatibility, and memory management for safe hand-off to Python and external consumers. [[1]](diffhunk://#diff-6f8d201a6a0399da3354d935e58302f437007e3dc59041f9e3a3010b373c05aeR1-R278) [[2]](diffhunk://#diff-98009d538063c4d6228fe281f355afdc793908bb750e1c650c980937ff924b48R8)

These changes together enable seamless, high-performance integration of AIG network structures with modern ML pipelines and frameworks.

Resolves #307

## Checklist:

- [ ] The pull request only contains commits that are related to it.
- [ ] I have added appropriate tests and documentation.
- [ ] I have made sure that all CI jobs on GitHub pass.
- [ ] The pull request introduces no new warnings and follows the project's style guidelines.

## Summary by CodeRabbit

* **New Features**
  * Export AIG networks as graph tensors (edge_index, edge_attr, node_attr) with selectable node/edge encodings, optional level/fanout, and optional truth-table export; Python enums and to_graph_tensors method added. Sequential networks raise on unsupported export.

* **Documentation**
  * New DLPack Tensors guide with end-to-end, zero-copy examples for PyTorch/JAX/TF/NumPy, dtype/shape mappings, usage notes, and performance guidance.

* **Tests**
  * Comprehensive tests for encodings, shapes/dtypes, truth-table handling, DLPack interop, large graphs, and edge cases.

### CODE REVIEW

**Comment by marcelwa** [comment]
@copilot resolve the merge conflicts in this pull request
